### ISSUE TITLE
Figure out what happens during daemon startup and what can be improved

### ISSUE THREAD

**Issue by lyoshenka** [lbryio/grooming#18]

**Comment by hackrush01** [comment]
Some points based on preliminary discussion and analyses
 - Blacklist those methods which couldn't be called during startup as opposed to white-listing them
 - Reduce the blacklisted method list as we progress in the daemon's startup process(for e.g. wallet commands should be available after wallet startup finishes)

Update:
 - Consider using decorators.

**Comment by kauffj** [comment]
Perhaps rather than blacklisting, each sub-component could register the API functions it provides once it is ready/loaded?

**Comment by hackrush01** [comment]
As per discussion we figured that a blacklisted list would be smaller. But this is still open for discussion, and "the registering the API functions" solution would certainly require less overall changes in the daemon(as part of it is already implemented using the `allowed_during_startup` list).
But if there are only a handful of functions that aren't allowed during startup, then blacklisting would make more sense IMO.
(Added a new update the above list as well)

**Comment by kauffj** [comment]
@jackrobison told me he anticipates the daemon shifting to a model where we would build binaries with and without different subcomponents (e.g. building the daemon without the data network for standalone wallet purposes, or building the daemon without a wallet for reflector purposes).

It may be worth discussing with him briefly so we can avoid redesigning this again in several months.

**Comment by hackrush01** [comment]
Killing two stones with one bird here actually
Currently we are running almost everything as deferreds during the daemon startup, so it can make sense to run some/most of them as a different thread so that each startup phase can become non-blocking, at least all the deferreds which do not require chaining or are sequential can be run this way.
Thought of this while working on lbryio/lbryum#175 and reading up on the Daemon's code.

**Comment by eukreign** [comment]
Please don't use threads, you can run many deferreds in parallel by `yield`ing on a [defer.DeferredList](https://twistedmatrix.com/documents/16.5.0/core/howto/defer.html#deferredlist).

**Comment by hackrush01** [comment]
Try to batch calls which query network/blockchain or make them non-blocking by running them after the required parts of daemon have started up. For instance checking up if all the blobs are present in reflector can either be batched or done after the daemon startup

**Comment by hackrush01** [comment]
@eukreign I read the article about DeferredList but to me it doesn't seem that it runs it in truly parallel fashion. It just seems to run all the deferred sequentially and give the result in a list fashion. Did I not understand something correctly?
Also, why is not advised to use thread to offer the required parallelism?

**Comment by eukreign** [comment]
To use threads the code has to be thread safe. For the most part the code is not thread safe, except for a few locks here and there. You'd have to be a masochist to try and make a code base as large as lbry completely thread safe. A lot of the intermittent bugs and issue we are having could be due to all deferToThread calls. 

https://en.m.wikipedia.org/wiki/Thread_safety

http://wiki.c2.com/?ThreadsConsideredHarmful

https://web.stanford.edu/~ouster/cgi-bin/papers/threads.pdf





DeferredList runs in parallel, that's the whole point of it.

**Comment by eukreign** [comment]
@hackrush01 maybe this will illustrate how it works:

```python
from __future__ import print_function
import time
from twisted.internet import task, defer


def two_sec_task(reactor, name):
    d = defer.Deferred()
    d.addCallback(print)
    reactor.callLater(2, d.callback, name+' done')
    return d


@defer.inlineCallbacks
def sequential(reactor):
    print('\nsequential:')
    start = time.time()
    yield two_sec_task(reactor, 'first')
    yield two_sec_task(reactor, 'second')
    yield two_sec_task(reactor, 'third')
    print('Elapsed: {}s'.format(time.time()-start))


@defer.inlineCallbacks
def parallel_deferredlist(reactor):
    print('\nparallel using DeferredList:')
    start = time.time()
    yield defer.DeferredList([
      two_sec_task(reactor, 'first'),
      two_sec_task(reactor, 'second'),
      two_sec_task(reactor, 'third')
    ])
    print('Elapsed: {}s'.format(time.time()-start))


@defer.inlineCallbacks
def main(reactor):
    yield sequential(reactor)
    yield parallel_deferredlist(reactor)


task.react(main)
```

This is the output:
```
sequential:
first done
second done
third done
Elapsed: 6.00413203239s

parallel using DeferredList:
first done
second done
third done
Elapsed: 2.00023603439s
```

`sequential()` took 6 seconds to run because each task had to run after the next task. in the DefferredList version they all ran at the same time and so the total time to run them all was 2 seconds.

As an exercise you can try and implement a version of parallel without using DeferredList, you can look at how DeferredList is implemented for inspiration.

**Comment by hackrush01** [comment]
Thanks a lot @eukreign. The example was really helpful, clear and to the point.

**Comment by hackrush01** [comment]
Broken them down in #27 #28 #29
Issue epic in lbryio/lbry#1121
